samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -59,6 +61,7 @@
 @InterfaceStability.Evolving
 public class StreamProcessor {
   private final JobCoordinator jobCoordinator;
+  private final List<StreamProcessorLifeCycleAware> lifeCycleAwares = new ArrayList<>();
```

Comment:
Why is this a list of StreamProcessorLifeCycleAware? 
Iirc, there is a 1:N mapping between an ApplicationRunner and StreamProcessor. Shouldn't this relationship be maintained on the ApplicationRunner layer, as opposed to StreamProcessor layer? 
It is extremely weird to see StreamProcessor taking in multiple listeners. If the relationship were inverted where there are N ApplicationRunners and we need to notify each of them from a single StreamProcessor, this makes sense. But that is not the case. 
Can you please explain the motivation behind this choice?